src/modules/newtek/producer/newtek_ndi_producer.cpp

```diff
@@ -0,0 +1,305 @@
+    const int          instance_no_;
+    const std::wstring name_;
+    const bool         low_bandwidth_;
+
+    spl::shared_ptr<core::frame_factory> frame_factory_;
+    core::video_format_desc              format_desc_;
+    NDIlib_v3*                           ndi_lib_;
+    NDIlib_framesync_instance_t          ndi_framesync_;
+    NDIlib_recv_instance_t               ndi_recv_instance_;
+    NDIlib_video_frame_v2_t              ndi_video_frame_;
+    NDIlib_audio_frame_v2_t              ndi_audio_frame_;
+    spl::shared_ptr<diagnostics::graph>  graph_;
+    timer                                tick_timer_;
+    timer                                frame_timer_;
+
+    std::queue<core::draw_frame> frames_;
+    mutable std::mutex           frames_mutex_;
+    core::draw_frame             last_frame_ = core::draw_frame::empty();
+    int                          frame_no_;
+    executor                     executor_;
+
+    int cadence_counter_;
+    int cadence_length_;
+
+  public:
+    explicit newtek_ndi_producer(spl::shared_ptr<core::frame_factory> frame_factory,
+                                 core::video_format_desc              format_desc,
+                                 std::wstring                         name,
+                                 bool                                 low_bandwidth)
+        : format_desc_(format_desc)
+        , frame_factory_(frame_factory)
+        , name_(name)
+        , low_bandwidth_(low_bandwidth)
+        , executor_(print())
+        , instance_no_(instances_++)
+        , cadence_counter_(0)
+        , frame_no_(0)
+    {
+        if (!ndi::load_library()) {
+            ndi::not_installed();
+        }
+
+        graph_->set_text(print());
+        graph_->set_color("frame-time", diagnostics::color(0.5f, 1.0f, 0.2f));
+        graph_->set_color("tick-time", diagnostics::color(0.0f, 0.6f, 0.9f));
+        graph_->set_color("dropped-frame", diagnostics::color(0.3f, 0.6f, 0.3f));
+        diagnostics::register_graph(graph_);
+        executor_.set_capacity(1);
+        cadence_length_ = static_cast<int>(format_desc_.audio_cadence.size());
+        initialize();
+    }
+
+    ~newtek_ndi_producer()
+    {
+        executor_.stop();
+        ndi_lib_->NDIlib_framesync_destroy(ndi_framesync_);
+        ndi_lib_->NDIlib_recv_destroy(ndi_recv_instance_);
+    }
+
+    std::wstring print() const override
+    {
+        return L"ndi[" + boost::lexical_cast<std::wstring>(instance_no_) + L"|" + name_ + L"]";
+    }
+
+    std::wstring name() const override { return L"ndi"; }
+
+    core::draw_frame receive_impl(int hints) override
+    {
+        graph_->set_value("tick-time", tick_timer_.elapsed() * format_desc_.fps * 0.5);
+        tick_timer_.restart();
+        core::draw_frame frame;
+        if (0 == frame_no_++) {
+            executor_.invoke([this]() { prepare_next_frame(); });
+        }
+        executor_.begin_invoke([this]() { prepare_next_frame(); });
```

Comment:
The idea behind using FrameSync is that `NDIlib_framesync_capture_video` and `NDIlib_framesync_capture_audio` return immediately, as all the necessary buffering(and repeating in case of underflow) is done by the library itself. The only thing that may take some time in `prepare_next_frame` is calling `ffmpeg::make_frame`, but still it shouldn't take longer than a frame. I did it this way so that calls to `receive_impl` could clock receiving a/v from the FrameSync. It skips the need to implement any buffering and audio resampling. `executor_.invoke` for the first frame is called to prevent a blank frame appearing when doing `LOADBG` and then `PLAY`, and I can imagine this messing some things up, but I have yet to look closer at graphs and performance because so far I haven't seen any anomalies.

Reply:
It should be OK to return `core::draw_frame{}` before any data is ready, and the layer will keep the previous producer alive until the new one returns a real frame. I believe this is how the ffmpeg producer works, to avoid blocking mixing like this.

OK. Even though it should take very little time, I still think it would be good to not have a chance it could block.
A simple option would be to allow executor_ to have a longer queue, and don't schedule another if there are a couple waiting to execute. That keeps the timing controlled by receive, with no risk of blocking

Reply:
Oh, I see now. The black frame showed up because I returned core::draw_frame::empty().

Reply:
I set executor_'s capacity to 2 and schedule preparing new frame if there is no more than one in its queue. Is it fine?